// weaviate-agents/dist/query/response/response.js
export var ComparisonOperator;
(function (ComparisonOperator) {
    ComparisonOperator["Equals"] = "=";
    ComparisonOperator["LessThan"] = "<";
    ComparisonOperator["GreaterThan"] = ">";
    ComparisonOperator["LessEqual"] = "<=";
    ComparisonOperator["GreaterEqual"] = ">=";
    ComparisonOperator["NotEquals"] = "!=";
    ComparisonOperator["Like"] = "LIKE";
})(ComparisonOperator || (ComparisonOperator = {}));
export var NumericMetrics;
(function (NumericMetrics) {
    NumericMetrics["Count"] = "COUNT";
    NumericMetrics["Maximum"] = "MAXIMUM";
    NumericMetrics["Mean"] = "MEAN";
    NumericMetrics["Median"] = "MEDIAN";
    NumericMetrics["Minimum"] = "MINIMUM";
    NumericMetrics["Mode"] = "MODE";
    NumericMetrics["Sum"] = "SUM";
    NumericMetrics["Type"] = "TYPE";
})(NumericMetrics || (NumericMetrics = {}));
export var TextMetrics;
(function (TextMetrics) {
    TextMetrics["Count"] = "COUNT";
    TextMetrics["Type"] = "TYPE";
    TextMetrics["TopOccurrences"] = "TOP_OCCURRENCES";
})(TextMetrics || (TextMetrics = {}));
export var BooleanMetrics;
(function (BooleanMetrics) {
    BooleanMetrics["Count"] = "COUNT";
    BooleanMetrics["Type"] = "TYPE";
    BooleanMetrics["TotalTrue"] = "TOTAL_TRUE";
    BooleanMetrics["TotalFalse"] = "TOTAL_FALSE";
    BooleanMetrics["PercentageTrue"] = "PERCENTAGE_TRUE";
    BooleanMetrics["PercentageFalse"] = "PERCENTAGE_FALSE";
})(BooleanMetrics || (BooleanMetrics = {}));
export var DateMetrics;
(function (DateMetrics) {
    DateMetrics["Count"] = "COUNT";
    DateMetrics["Maximum"] = "MAXIMUM";
    DateMetrics["Median"] = "MEDIAN";
    DateMetrics["Minimum"] = "MINIMUM";
    DateMetrics["Mode"] = "MODE";
})(DateMetrics || (DateMetrics = {}));
